packages/tpetra/core/src/Tpetra_Distributor.hpp

```diff
@@ -2306,6 +2549,104 @@ namespace Tpetra {
+
+              Packet t;
+              MPI_Datatype rawType = Teuchos::Details::MpiTypeTraits<Packet>::getType (t);
+              Packet* rawRecvBuf = recvBuf.data ();
+              const int count = as<int> (recvBuf.size ());
+              MPI_Request rawRequest = MPI_REQUEST_NULL;
+              MPI_Recv_init (rawRecvBuf, count, rawType, procsFrom_[i], tag, rawComm, &rawRequest);
+
+              Teuchos::RCP<Teuchos::CommRequest<int> > req = Teuchos::mpiCommRequest<int>(rawRequest, count);
+              // Teuchos::set_extra_data( globalRecvBuf, "buffer", inOutArg(req) );
+              persistentRequests_.push_back(req);
+            }
+            else { // Receiving from myself
+              selfReceiveOffset = curBufOffset; // Remember the self-recv offset
+            }
+            curBufOffset += curBufLen;
+          }
+
+          if (verbose_) {
+            std::ostringstream os;
+            os << "Proc " << myRank << ": Creating persistent send requests" << endl;
+            *out_ << os.str ();
+          }
+          // create the persistent send requests
+          for (size_t i = 0; i < numBlocks; ++i) {
+            size_t p = i + procIndex;
+            if (p > (numBlocks - 1)) {
+              p -= numBlocks;
+            }
+
+            if (procsTo_[p] != myRank) {
+
+              exports_view_type sendBuf =
+                subview_offset (exports, startsTo_[p] * numPackets,
+                                lengthsTo_[p] * numPackets);
+
+              Packet t;
+              MPI_Datatype rawType = Teuchos::Details::MpiTypeTraits<Packet>::getType (t);
+              const Packet* rawSendBuf = sendBuf.data ();
+              const int count = as<int> (sendBuf.size ());
+              MPI_Request rawRequest = MPI_REQUEST_NULL;
+              MPI_Send_init (rawSendBuf, count, rawType, procsTo_[p], tag, rawComm, &rawRequest);
+
+              Teuchos::RCP<Teuchos::CommRequest<int> > req = Teuchos::mpiCommRequest<int>(rawRequest, count);
+              // Teuchos::set_extra_data( sendBuf, "buffer", inOutArg(req) );
+              persistentRequests_.push_back(req);
+            }
+          }
+        }
+
+        if (verbose_) {
+          std::ostringstream os;
+          os << "Proc " << myRank << ": Starting persistent requests" << endl;
+          *out_ << os.str ();
+        }
+        int count = persistentRequests_.size();
+        Array<MPI_Request> rawMpiRequests (count, MPI_REQUEST_NULL);
+        int k = 0;
+        for (int i = 0; i<persistentRequests_.size(); i++) {
+          Teuchos::RCP<Teuchos::CommRequest<int> > request = persistentRequests_[i];
+          Teuchos::RCP<Teuchos::MpiCommRequestBase<int> > mpiRequest =
+            Teuchos::rcp_dynamic_cast<Teuchos::MpiCommRequestBase<int> > (request);
+          rawMpiRequests[k] = mpiRequest->getRawMpiRequest();
+          k++;
+        }
+        const int err = MPI_Startall (count, rawMpiRequests.getRawPtr());
```

Comment:
See comment above.  Why do we use `MPI_Startall` instead of starting each message separately?

Reply:
It seemed cleaner to me, and maybe there could be some MPI optimization benefits?

Reply:
Ah, sorry, never mind.  (The idea is that you have the requests all set up already, so of course you want to start them all at once.)